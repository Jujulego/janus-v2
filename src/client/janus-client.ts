import { TypedDocumentNode } from '@graphql-typed-document-node/core';
import { Logger, logger$, withLabel } from '@kyrielle/logger';
import { DocumentNode, FormattedExecutionResult, OperationDefinitionNode, print } from 'graphql';
import { Client, createClient, ExecutionResult, RequestParams } from 'graphql-sse';
import { AsyncReadable, type Observable, observable$ } from 'kyrielle';
import assert from 'node:assert';

import { health$, HealthPayload } from './health.ref.js';

// Types
export interface OperationOptions {
  readonly variables?: Record<string, unknown>;
  readonly signal?: AbortSignal;
}

// Class
export class JanusClient implements Disposable {
  // Attributes
  private _sseClient: Client<false> | null = null;
  private _healthController = new AbortController();

  readonly logger: Logger;
  readonly serverHealth$: AsyncReadable<HealthPayload>;
  readonly [Symbol.dispose]: () => void;

  // Constructor
  constructor(
    readonly janusUrl = 'http://localhost:3000/',
    logger = logger$(),
  ) {
    this.logger = logger.child(withLabel('janus'));

    this.serverHealth$ = health$(
      new URL('/_janus/health', janusUrl),
      this.logger.child(withLabel('janus:health')),
    );

    this[Symbol.dispose ?? Symbol.for('Symbol.dispose')] = this.dispose.bind(this);
  }

  // Methods
  private _prepareQuery(document: DocumentNode, variables?: Record<string, unknown>): RequestParams {
    const operation = document.definitions
      .find((def): def is OperationDefinitionNode => def.kind === 'OperationDefinition');

    const request: RequestParams = { query: print(document) };

    if (operation?.name?.value) {
      request.operationName = operation.name.value;
    }

    if (variables) {
      request.variables = variables;
    }

    return request;
  }

  /**
   * Connects client to server, and setup healthcheck.
   * @returns `true` if connection is successful, `false` otherwise.
   */
  async initiate(signal?: AbortSignal): Promise<HealthPayload> {
    const health = await this.serverHealth$.read(signal);

    if (!this._sseClient) {
      this.logger.verbose`Janus client connected (server version: ${health.version})`;

      const url = new URL('/_janus/graphql', this.janusUrl).toString();
      this.logger.debug`Initiate sse client against ${url}`;

      this._sseClient = createClient({
        url,
        retry: async () => {
          await this.serverHealth$.read(this._healthController.signal);
        },
        on: {
          connecting: () => this.logger.debug`Connecting to sse stream`,
          message: (message) => this.logger.debug`Sending ${message.event} to sse stream`,
          connected: () => this.logger.debug`Connected to sse stream`,
        }
      });
    }

    return health;
  }

  /**
   * Send query to the server
   */
  async send<R, V>(document: TypedDocumentNode<R, V>, opts: OperationOptions = {}): Promise<FormattedExecutionResult<R>> {
    const query = this._prepareQuery(document, opts.variables);
    this.logger.debug`Sending ${query.operationName ?? 'graphql'} request to server at ${this.janusUrl}`;

    const res = await fetch(new URL('/_janus/graphql', this.janusUrl), {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json; charset=utf-8'
      },
      body: JSON.stringify(query),
      signal: opts.signal ?? null,
    });

    return await res.json();
  }

  /**
   * Subscribes to an event stream
   */
  subscribe$<D>(document: TypedDocumentNode<D, Record<string, never>>): Observable<ExecutionResult<D>>;
  subscribe$<D, V extends Record<string, unknown>>(document: TypedDocumentNode<D, V>, variables: V): Observable<ExecutionResult<D>>;
  subscribe$<D, V extends Record<string, unknown>>(document: TypedDocumentNode<D, V>, variables?: V): Observable<ExecutionResult<D>> {
    assert(!!this._sseClient, 'Client should be initiated before any observe call');

    return observable$<ExecutionResult<D>>((observer, signal) => {
      const off = this._sseClient!.subscribe<D>(this._prepareQuery(document, variables), observer);
      signal.addEventListener('abort', off, { once: true });
    });
  }

  /**
   * Dispose all resources connected with the server
   */
  dispose(): void {
    this._healthController.abort();

    if (this._sseClient) {
      this._sseClient.dispose();
      this._sseClient = null;

      this.logger.verbose`Janus client disconnected`;
    }
  }
}
