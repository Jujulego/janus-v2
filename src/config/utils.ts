import { inject$, token$ } from '@jujulego/injector';
import { cosmiconfig, defaultLoaders } from 'cosmiconfig';

import { Ajv } from '../ajv.config.js';
import { dynamicImport } from '../utils/import.js';
import { Config } from './type.js';

// Schema
const schema = {
  $schema: 'http://json-schema.org/draft-07/schema',
  type: 'object',
  additionalProperties: false,
  required: ['redirections'],
  properties: {
    pidfile: {
      type: 'string',
      default: '.janus.pid'
    },
    proxy: {
      type: 'object',
      additionalProperties: false,
      properties: {
        port: {
          type: 'number',
          default: 3000
        }
      }
    },
    redirections: {
      type: 'array',
      items: {
        $ref: '#/definitions/redirection'
      },
      minItems: 1
    }
  },
  definitions: {
    output: {
      type: 'object',
      additionalProperties: false,
      required: ['target'],
      properties: {
        target: {
          type: 'string'
        },
        enabled: {
          type: 'boolean',
          default: false
        },
        changeOrigin: {
          type: 'boolean',
          default: false
        },
        secure: {
          type: 'boolean',
          default: false
        },
        ws: {
          type: 'boolean',
          default: false
        }
      }
    },
    redirection: {
      type: 'object',
      additionalProperties: false,
      required: ['url', 'outputs'],
      properties: {
        url: {
          type: 'string'
        },
        outputs: {
          type: 'object',
          minProperties: 1,
          additionalProperties: {
            $ref: '#/definitions/output'
          }
        }
      }
    }
  }
};

// Tokens
export const ConfigValidator = token$(
  () => inject$(Ajv).compile<Config>(schema)
);

export const ConfigExplorer = token$(
  () => cosmiconfig('janus', {
    loaders: {
      '.cjs': (filepath) => dynamicImport(filepath).then((mod) => mod.default),
      '.js': (filepath) => dynamicImport(filepath).then((mod) => mod.default),
      '.json': defaultLoaders['.json'],
      '.yaml': defaultLoaders['.yaml'],
      '.yml': defaultLoaders['.yml'],
      noExt: defaultLoaders.noExt,
    }
  })
);
