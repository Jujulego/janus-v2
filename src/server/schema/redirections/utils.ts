import { RedirectionState } from '../../store/redirections/types.js';
import { Redirection } from '../schema.types.js';

// Utils
export function mapRedirection(redirection: RedirectionState | null): Redirection | null {
  return redirection && {
    id: redirection.id,
    url: redirection.url,
    outputs: redirection.outputs.map((name) => redirection.outputsByName[name]!)
  };
}

export function mapRedirectionList(list: RedirectionState[]): Redirection[] {
  return list.map((r) => mapRedirection(r)!);
}
