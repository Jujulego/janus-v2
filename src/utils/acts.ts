import { SyncMutableRef } from '@jujulego/aegis';
import { produce } from 'immer';

// Types
export type ActReducer<A extends unknown[] = unknown[], T = unknown> = (...args: A) => (state: T) => void;
export type Act<A extends unknown[] = unknown[], T = unknown> = (...args: A) => T;

export type ActRef<T, A extends Record<string, ActReducer<unknown[], T>>> = {
  [K in keyof A]: A[K] extends ActReducer<infer Args, T> ? Act<Args, T> : never;
} & SyncMutableRef<T>;

export function acts<T, A extends Record<string, ActReducer<any[], T>>>(ref: SyncMutableRef<T>, acts: A): ActRef<T, A> {
  const result: SyncMutableRef<T> = {
    next: ref.next,
    read: ref.read,
    mutate: ref.mutate,
    subscribe: ref.subscribe,
    unsubscribe: ref.unsubscribe,
    clear: ref.clear,
  };

  for (const [key, act] of Object.entries(acts)) {
    Object.assign(result, {
      [key]: (...args: any[]) => ref.mutate(produce(ref.read(), act(...args))),
    });
  }

  return result as ActRef<T, A>;
}