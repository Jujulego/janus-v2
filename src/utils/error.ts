export function isTimeoutError(err: unknown): err is DOMException {
  return err instanceof DOMException && ['AbortError', 'TimeoutError'].includes(err.name);
}