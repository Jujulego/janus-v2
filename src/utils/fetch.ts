import { Readable, readable$ } from 'kyrielle';

export interface FetchOpts extends Omit<RequestInit, 'signal'> {
  onFetch?: (url: string | URL) => void;
}

export class FetchError extends Error {
  // Constructor
  constructor(readonly response: Response) {
    super(response.statusText);
  }

  // Properties
  get status(): number {
    return this.response.status;
  }
}

export function fetch$(url: string | URL, opts: FetchOpts = {}): Readable<Promise<unknown>> {
  return readable$(async (signal): Promise<unknown> => {
    if (opts.onFetch) opts.onFetch(url);
    const res = await fetch(url, { ...opts, signal });

    if (!res.ok) {
      throw new FetchError(res);
    }

    return res.json();
  });
}
