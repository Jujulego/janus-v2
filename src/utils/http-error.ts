import { HttpError } from 'http-errors';
import { ServerResponse } from 'node:http';

/**
 * Send http error to response
 */
export function sendHttpError(res: ServerResponse, error: HttpError) {
  res.statusCode = error.statusCode;
  res.setHeader('Content-Type', 'application/json');
  res.write(JSON.stringify({
    status: error.statusCode,
    message: error.message
  }));
  res.end();
}

/**
 * Render as a valid http response
 */
export function renderHttpError(error: HttpError): string {
  const content = JSON.stringify({
    status: error.statusCode,
    message: error.message
  });

  return `HTTP/1.1 ${error.statusCode} ${error.name}\r\n`
       + `Content-Length: ${content.length}\r\n`
       + 'Content-Type: application/json\r\n'
       + '\r\n'
       + '\r\n'
       + content;
}
