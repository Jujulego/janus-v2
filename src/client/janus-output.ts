import { logger$ } from '@kyrielle/logger';
import { pipe$, retry$, timeout$ } from 'kyrielle';
import assert from 'node:assert';

import { ConfigService } from '../config/config.service.js';
import { Config, OutputConfig, RedirectionConfig } from '../config/type.js';
import { generateRedirectionId } from '../utils/redirections.js';
import { graphql } from './gql/index.js';
import { JanusClient } from './janus-client.js';

// Types
export interface JanusOutputOptions {
  configFile?: string;
}

/**
 * Manages a given output.
 */
export class JanusOutput implements Disposable {
  // Attributes
  readonly logger = logger$();
  readonly redirectionId: string;
  readonly [Symbol.dispose]: () => void;

  private _client?: JanusClient;
  private _config?: Config;

  private readonly _configService = new ConfigService(this.logger);

  // Constructor
  constructor(
    readonly redirectionUrl: string,
    readonly outputName: string,
    readonly options: JanusOutputOptions = {},
  ) {
    this.redirectionId = generateRedirectionId(this.redirectionUrl);

    this[Symbol.dispose ?? Symbol.for('Symbol.dispose')] = this.dispose.bind(this);
  }

  // Methods
  private async _getConfig(): Promise<Config> {
    if (!this._config) {
      if (this.options?.configFile) {
        this._config = await this._configService.loadConfig(this.options.configFile);
      } else {
        this._config = await this._configService.searchConfig();
      }
    }

    return this._config;
  }

  private async _getClient(): Promise<JanusClient> {
    if (!this._client) {
      const config = await this._getConfig();
      this._client = new JanusClient(`http://localhost:${config.server.port}/`, this.logger);
    }

    return this._client;
  }

  private async _getRedirectionOutput(): Promise<{ redirection: RedirectionConfig, output: OutputConfig }> {
    const config = await this._getConfig();

    // Search redirection
    const redirection = config.redirections[this.redirectionUrl];

    if (!redirection) {
      throw new Error(`No redirection configured at ${this.redirectionUrl}`);
    }

    // Search output
    const output = redirection.outputs[this.outputName];

    if (!output) {
      throw new Error(`No output ${this.outputName} found at ${this.redirectionUrl}`);
    }

    return { redirection, output };
  }

  /**
   * Enables output in server
   * @param signal
   */
  async enable(signal?: AbortSignal): Promise<void> {
    await this._getRedirectionOutput(); // <= asserts that output exists according to config
    const client = await this._getClient();

    this.logger.info('Enabling redirection output...');
    const { data: result } = await pipe$(
      client.request$(
        graphql(`
          mutation EnableJanusOutput($redirectionId: ID!, $outputName: String!) {
            enableRedirectionOutput(redirectionId: $redirectionId, outputName: $outputName) {
              id
            }
          }
        `),
        { redirectionId: this.redirectionId, outputName: this.outputName }
      ),
      retry$('read', {
        tryTimeout: 1000,
        onRetry: () => timeout$(1000),
      }),
    ).read(signal);

    assert(result?.enableRedirectionOutput?.id === this.redirectionId, 'Janus server responded with wrong redirection id');
  }

  /**
   * Disables output in server
   * @param signal
   */
  async disable(signal?: AbortSignal): Promise<void> {
    await this._getRedirectionOutput(); // <= asserts that output exists according to config
    const client = await this._getClient();

    this.logger.info('Disabling redirection output...');
    const { data: result } = await pipe$(
      client.request$(
        graphql(`
          mutation DisableJanusOutput($redirectionId: ID!, $outputName: String!) {
            disableRedirectionOutput(redirectionId: $redirectionId, outputName: $outputName) {
              id
            }
          }
        `),
        { redirectionId: this.redirectionId, outputName: this.outputName }
      ),
      retry$('read', {
        tryTimeout: 1000,
        onRetry: () => timeout$(1000),
      }),
    ).read(signal);

    assert(result?.disableRedirectionOutput?.id === this.redirectionId, 'Janus server responded with wrong redirection id');
  }

  /**
   * Disposes internal resources
   */
  dispose(): void {
    this._client?.dispose();
    delete this._client;
  }
}
