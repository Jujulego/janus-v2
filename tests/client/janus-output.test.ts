import { graphql as mockGql, HttpResponse } from 'msw';
import { setupServer } from 'msw/node';
import { afterAll, afterEach, beforeAll, beforeEach, describe, expect, it, vi } from 'vitest';

import { JanusClient } from '@/src/client/janus-client.js';
import { JanusOutput } from '@/src/client/janus-output.js';
import { ConfigService } from '@/src/config/config.service.js';
import { DEFAULT_CONFIG } from '@/tests/utils.js';

// Constants
const config = {
  ...DEFAULT_CONFIG,
  redirections: {
    '/life': {
      outputs: {
        book: {
          target: 'http://localhost:3042',
          enabled: true,
          changeOrigin: false,
          secure: false,
          ws: false,
        },
      },
    },
  }
};

// Mock server
const janusGql = mockGql.link(`http://localhost:${config.server.port}/_janus/graphql`);

const mockServer = setupServer(
  janusGql.mutation('EnableJanusOutput', ({ variables }) => HttpResponse.json({
    data: {
      enableRedirectionOutput: { id: variables.redirectionId }
    }
  })),
  janusGql.mutation('DisableJanusOutput', ({ variables }) => HttpResponse.json({
    data: {
      disableRedirectionOutput: { id: variables.redirectionId }
    }
  }))
);

// Setup
let output: JanusOutput;

beforeAll(() => {
  mockServer.listen();
});

beforeEach(() => {
  mockServer.resetHandlers();

  output = new JanusOutput('/life', 'book');

  vi.spyOn(ConfigService.prototype, 'loadConfig').mockResolvedValue(config);
  vi.spyOn(ConfigService.prototype, 'searchConfig').mockResolvedValue(config);

  vi.spyOn(JanusClient.prototype, 'request$');
});

afterEach(async () => {
  output?.dispose();
});

afterAll(() => {
  mockServer.close();
});

// Tests
describe('JanusOutput.enable', () => {
  it('should send enable query to server according to config', async () => {
    await output.enable();

    // Config loading
    expect(ConfigService.prototype.loadConfig).not.toHaveBeenCalled();
    expect(ConfigService.prototype.searchConfig).toHaveBeenCalled();

    // Graphql call
    expect(JanusClient.prototype.request$).toHaveBeenCalledOnce();
    expect(vi.mocked(JanusClient.prototype.request$).mock.calls[0]).toMatchSnapshot();
  });

  it('should load given config file', async () => {
    const output = new JanusOutput('/life', 'book', { configFile: '/janus.yml' });
    await output.enable();

    // Config loading
    expect(ConfigService.prototype.loadConfig).toHaveBeenCalledWith('/janus.yml');
    expect(ConfigService.prototype.searchConfig).not.toHaveBeenCalled();

    // Graphql call
    expect(JanusClient.prototype.request$).toHaveBeenCalledOnce();
    expect(vi.mocked(JanusClient.prototype.request$).mock.calls[0]).toMatchSnapshot();
  });

  it('should throw as given redirection does not exists', async () => {
    const output = new JanusOutput('/toto', 'book', { configFile: '/janus.yml' });
    await expect(output.enable())
      .rejects.toEqual(new Error('No redirection configured at /toto'));

    // Graphql call
    expect(JanusClient.prototype.request$).not.toHaveBeenCalled();
  });

  it('should throw as given output does not exists', async () => {
    const output = new JanusOutput('/life', 'toto', { configFile: '/janus.yml' });
    await expect(output.enable())
      .rejects.toEqual(new Error('No output toto found at /life'));

    // Graphql call
    expect(JanusClient.prototype.request$).not.toHaveBeenCalled();
  });
});

describe('JanusOutput.disable', () => {
  it('should send disable query to server according to config', async () => {
    await output.disable();

    // Config loading
    expect(ConfigService.prototype.loadConfig).not.toHaveBeenCalled();
    expect(ConfigService.prototype.searchConfig).toHaveBeenCalled();

    // Graphql call
    expect(JanusClient.prototype.request$).toHaveBeenCalledOnce();
    expect(vi.mocked(JanusClient.prototype.request$).mock.calls[0]).toMatchSnapshot();
  });

  it('should load given config file', async () => {
    const output = new JanusOutput('/life', 'book', { configFile: '/janus.yml' });
    await output.disable();

    // Config loading
    expect(ConfigService.prototype.loadConfig).toHaveBeenCalledWith('/janus.yml');
    expect(ConfigService.prototype.searchConfig).not.toHaveBeenCalled();

    // Graphql call
    expect(JanusClient.prototype.request$).toHaveBeenCalledOnce();
    expect(vi.mocked(JanusClient.prototype.request$).mock.calls[0]).toMatchSnapshot();
  });

  it('should throw as given redirection does not exists', async () => {
    const output = new JanusOutput('/toto', 'book', { configFile: '/janus.yml' });
    await expect(output.disable())
      .rejects.toEqual(new Error('No redirection configured at /toto'));

    // Graphql call
    expect(JanusClient.prototype.request$).not.toHaveBeenCalled();
  });

  it('should throw as given output does not exists', async () => {
    const output = new JanusOutput('/life', 'toto', { configFile: '/janus.yml' });
    await expect(output.disable())
      .rejects.toEqual(new Error('No output toto found at /life'));

    // Graphql call
    expect(JanusClient.prototype.request$).not.toHaveBeenCalled();
  });
});