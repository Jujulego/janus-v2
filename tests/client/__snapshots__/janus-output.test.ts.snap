// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JanusOutput.disable > should load given config file 1`] = `
[
  {
    "definitions": [
      {
        "kind": "OperationDefinition",
        "name": {
          "kind": "Name",
          "value": "DisableJanusOutput",
        },
        "operation": "mutation",
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [
            {
              "arguments": [
                {
                  "kind": "Argument",
                  "name": {
                    "kind": "Name",
                    "value": "redirectionId",
                  },
                  "value": {
                    "kind": "Variable",
                    "name": {
                      "kind": "Name",
                      "value": "redirectionId",
                    },
                  },
                },
                {
                  "kind": "Argument",
                  "name": {
                    "kind": "Name",
                    "value": "outputName",
                  },
                  "value": {
                    "kind": "Variable",
                    "name": {
                      "kind": "Name",
                      "value": "outputName",
                    },
                  },
                },
              ],
              "kind": "Field",
              "name": {
                "kind": "Name",
                "value": "disableRedirectionOutput",
              },
              "selectionSet": {
                "kind": "SelectionSet",
                "selections": [
                  {
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "id",
                    },
                  },
                ],
              },
            },
          ],
        },
        "variableDefinitions": [
          {
            "kind": "VariableDefinition",
            "type": {
              "kind": "NonNullType",
              "type": {
                "kind": "NamedType",
                "name": {
                  "kind": "Name",
                  "value": "ID",
                },
              },
            },
            "variable": {
              "kind": "Variable",
              "name": {
                "kind": "Name",
                "value": "redirectionId",
              },
            },
          },
          {
            "kind": "VariableDefinition",
            "type": {
              "kind": "NonNullType",
              "type": {
                "kind": "NamedType",
                "name": {
                  "kind": "Name",
                  "value": "String",
                },
              },
            },
            "variable": {
              "kind": "Variable",
              "name": {
                "kind": "Name",
                "value": "outputName",
              },
            },
          },
        ],
      },
    ],
    "kind": "Document",
  },
  {
    "outputName": "book",
    "redirectionId": "a26573220d20dd05d4a283582117eae3",
  },
]
`;

exports[`JanusOutput.disable > should send disable query to server according to config 1`] = `
[
  {
    "definitions": [
      {
        "kind": "OperationDefinition",
        "name": {
          "kind": "Name",
          "value": "DisableJanusOutput",
        },
        "operation": "mutation",
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [
            {
              "arguments": [
                {
                  "kind": "Argument",
                  "name": {
                    "kind": "Name",
                    "value": "redirectionId",
                  },
                  "value": {
                    "kind": "Variable",
                    "name": {
                      "kind": "Name",
                      "value": "redirectionId",
                    },
                  },
                },
                {
                  "kind": "Argument",
                  "name": {
                    "kind": "Name",
                    "value": "outputName",
                  },
                  "value": {
                    "kind": "Variable",
                    "name": {
                      "kind": "Name",
                      "value": "outputName",
                    },
                  },
                },
              ],
              "kind": "Field",
              "name": {
                "kind": "Name",
                "value": "disableRedirectionOutput",
              },
              "selectionSet": {
                "kind": "SelectionSet",
                "selections": [
                  {
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "id",
                    },
                  },
                ],
              },
            },
          ],
        },
        "variableDefinitions": [
          {
            "kind": "VariableDefinition",
            "type": {
              "kind": "NonNullType",
              "type": {
                "kind": "NamedType",
                "name": {
                  "kind": "Name",
                  "value": "ID",
                },
              },
            },
            "variable": {
              "kind": "Variable",
              "name": {
                "kind": "Name",
                "value": "redirectionId",
              },
            },
          },
          {
            "kind": "VariableDefinition",
            "type": {
              "kind": "NonNullType",
              "type": {
                "kind": "NamedType",
                "name": {
                  "kind": "Name",
                  "value": "String",
                },
              },
            },
            "variable": {
              "kind": "Variable",
              "name": {
                "kind": "Name",
                "value": "outputName",
              },
            },
          },
        ],
      },
    ],
    "kind": "Document",
  },
  {
    "outputName": "book",
    "redirectionId": "a26573220d20dd05d4a283582117eae3",
  },
]
`;

exports[`JanusOutput.enable > should load given config file 1`] = `
[
  {
    "definitions": [
      {
        "kind": "OperationDefinition",
        "name": {
          "kind": "Name",
          "value": "EnableJanusOutput",
        },
        "operation": "mutation",
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [
            {
              "arguments": [
                {
                  "kind": "Argument",
                  "name": {
                    "kind": "Name",
                    "value": "redirectionId",
                  },
                  "value": {
                    "kind": "Variable",
                    "name": {
                      "kind": "Name",
                      "value": "redirectionId",
                    },
                  },
                },
                {
                  "kind": "Argument",
                  "name": {
                    "kind": "Name",
                    "value": "outputName",
                  },
                  "value": {
                    "kind": "Variable",
                    "name": {
                      "kind": "Name",
                      "value": "outputName",
                    },
                  },
                },
              ],
              "kind": "Field",
              "name": {
                "kind": "Name",
                "value": "enableRedirectionOutput",
              },
              "selectionSet": {
                "kind": "SelectionSet",
                "selections": [
                  {
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "id",
                    },
                  },
                ],
              },
            },
          ],
        },
        "variableDefinitions": [
          {
            "kind": "VariableDefinition",
            "type": {
              "kind": "NonNullType",
              "type": {
                "kind": "NamedType",
                "name": {
                  "kind": "Name",
                  "value": "ID",
                },
              },
            },
            "variable": {
              "kind": "Variable",
              "name": {
                "kind": "Name",
                "value": "redirectionId",
              },
            },
          },
          {
            "kind": "VariableDefinition",
            "type": {
              "kind": "NonNullType",
              "type": {
                "kind": "NamedType",
                "name": {
                  "kind": "Name",
                  "value": "String",
                },
              },
            },
            "variable": {
              "kind": "Variable",
              "name": {
                "kind": "Name",
                "value": "outputName",
              },
            },
          },
        ],
      },
    ],
    "kind": "Document",
  },
  {
    "outputName": "book",
    "redirectionId": "a26573220d20dd05d4a283582117eae3",
  },
]
`;

exports[`JanusOutput.enable > should send enable query to server according to config 1`] = `
[
  {
    "definitions": [
      {
        "kind": "OperationDefinition",
        "name": {
          "kind": "Name",
          "value": "EnableJanusOutput",
        },
        "operation": "mutation",
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [
            {
              "arguments": [
                {
                  "kind": "Argument",
                  "name": {
                    "kind": "Name",
                    "value": "redirectionId",
                  },
                  "value": {
                    "kind": "Variable",
                    "name": {
                      "kind": "Name",
                      "value": "redirectionId",
                    },
                  },
                },
                {
                  "kind": "Argument",
                  "name": {
                    "kind": "Name",
                    "value": "outputName",
                  },
                  "value": {
                    "kind": "Variable",
                    "name": {
                      "kind": "Name",
                      "value": "outputName",
                    },
                  },
                },
              ],
              "kind": "Field",
              "name": {
                "kind": "Name",
                "value": "enableRedirectionOutput",
              },
              "selectionSet": {
                "kind": "SelectionSet",
                "selections": [
                  {
                    "kind": "Field",
                    "name": {
                      "kind": "Name",
                      "value": "id",
                    },
                  },
                ],
              },
            },
          ],
        },
        "variableDefinitions": [
          {
            "kind": "VariableDefinition",
            "type": {
              "kind": "NonNullType",
              "type": {
                "kind": "NamedType",
                "name": {
                  "kind": "Name",
                  "value": "ID",
                },
              },
            },
            "variable": {
              "kind": "Variable",
              "name": {
                "kind": "Name",
                "value": "redirectionId",
              },
            },
          },
          {
            "kind": "VariableDefinition",
            "type": {
              "kind": "NonNullType",
              "type": {
                "kind": "NamedType",
                "name": {
                  "kind": "Name",
                  "value": "String",
                },
              },
            },
            "variable": {
              "kind": "Variable",
              "name": {
                "kind": "Name",
                "value": "outputName",
              },
            },
          },
        ],
      },
    ],
    "kind": "Document",
  },
  {
    "outputName": "book",
    "redirectionId": "a26573220d20dd05d4a283582117eae3",
  },
]
`;
